server: ListTenants endpoint on system admin server

aadityasondhi opened the pull request:
This patch adds a new endpoint to the system admin server that returns a
list of active tenants in the cluster. It queries the `system.tenants`
table to enumerate a list of tenants and then it starts those tenants on
the host to be able to return SQL and RPC addresses for those tenants.

Fixes #95014.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95024)
<!-- Reviewable:end -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-:-NLWLzIgEVWpMPwLjc6R:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi)



<!-- Sent from Reviewable.io -->

aadityasondhi committed:
server: ListTenants endpoint on system admin server

This patch adds a new endpoint to the system admin server that returns a
list of active tenants in the cluster. It queries the `system.tenants`
table to enumerate a list of tenants and then it starts those tenants on
the host to be able to return SQL and RPC addresses for those tenants.

Fixes #95014.

Release note: None

aadityasondhi commented:
@dhartunian @knz Changed this to return rpc and sql addresses along with the name and ID. PTAL.

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi and @knz)

___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLaiWnL3-l4a742E6Nw):*
This makes me a bit nervous but is perhaps fine for the first iteration. If we're only running a single app tenant on the cluster and calling this endpoint just makes sure the tenant starts up on each node, that seems relatively benign. If the cluster is running many tenants I could see this creating problems.
___
*[`pkg/server/admin.go` line 4104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLajrsL6FxPaNWrOiqT:-NLajrsL6FxPaNWrOiqU:b-olf6d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d35a2c3931172db4e60865f5b03f108917b71188/pkg/server/admin.go#L4104)):*
> ```Go
> ) (*serverpb.ListTenantsResponse, error) {
> 	// Dummy import to pull in kvtenantccl. This allows us to start tenants.
> 	var _ = kvtenantccl.Connector{}
> ```

I don't think this is ok to do. `admin.go` can't pull in ccl code like this because it should be buildable as part of the OSS build.

See for instance how this code https://github.com/cockroachdb/cockroach/blob/a5ca1201e165d288ff188a6820843366028a920d/pkg/server/tenant.go#L828 works when CCL isn't present. Look at how the factory has a non-CCL implementation in a public `var` that gets overridden https://github.com/cockroachdb/cockroach/blob/50413ad4e7b87d4223f815858ae4655b179ad440/pkg/kv/kvclient/kvtenant/connector.go#L124
___
*[`pkg/server/admin.go` line 4133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLajaUy4v8VT63WoL5Y:-NLajaUy4v8VT63WoL5Z:b-kq770o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d35a2c3931172db4e60865f5b03f108917b71188/pkg/server/admin.go#L4133)):*
> ```Go
> 			return nil, err
> 		}
> 		if tenantServer, ok := server.(*tenantServerWrapper); ok {
> ```

instead of casting like this, consider implementing a method on the `onDemandServer` interface that gives you what you want. Then you can just call it in the loop without worrying about the tenant's special case.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
This is nice. Well done.

> Reviewed 2 of 5 files at r2, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-:-NLeRTHX07MCpJVPNEe9:b17ytcy)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi and @dhartunian)

___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLaiWnL3-l4a742E6Nw):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This makes me a bit nervous but is perhaps fine for the first iteration. If we're only running a single app tenant on the cluster and calling this endpoint just makes sure the tenant starts up on each node, that seems relatively benign. If the cluster is running many tenants I could see this creating problems.
</blockquote></details>

Yes this is why I foresee a cluster setting or other mechanism that lists the tenants to enumerate, instead of using the active field.
(We will also have a license-enforced limit on the number of running tenant servers btw)
___
*[`pkg/server/admin.go` line 4118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLeQwFn3J_jxqX_sbpy:-NLeQwFn3J_jxqX_sbpz:b-k71rav) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5e3eaf30740f5dd2662199d8f06ffb210f55e9/pkg/server/admin.go#L4118)):*
> ```Go
> 		tenantName, ok := tree.AsDString(row[0])
> 		if !ok {
> 			// If tenant name is NULL, we skip it because serverController will not support it.
> ```

Please add `WHERE ... AND name IS NOT NULL` above and get rid of this.
___
*[`pkg/server/admin.go` line 4129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLeR6WQ0RIzEgNai9d1:-NLeR6WQ0RIzEgNai9d2:b-be0pfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5e3eaf30740f5dd2662199d8f06ffb210f55e9/pkg/server/admin.go#L4129)):*
> ```Go
> 		server, err := s.server.serverController.getOrCreateServer(ctx, tenantName)
> 		if err != nil {
> 			return nil, err
> ```

Consider logging the error + skipping this server (or add an error field for this tenant name in the response payload), instead of aborting the list entirely. Remember that the zip logic was meant to do best effort collection - it's ok if it skips over things that don't work.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian and @knz)

___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLaiWnL3-l4a742E6Nw):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Yes this is why I foresee a cluster setting or other mechanism that lists the tenants to enumerate, instead of using the active field.
(We will also have a license-enforced limit on the number of running tenant servers btw)
</blockquote></details>

Ack. Nothing actionable re: this PR, right?
___
*[`pkg/server/admin.go` line 4104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLajrsL6FxPaNWrOiqT:-NM-O9iTDeo-S2AdMBSL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d35a2c3931172db4e60865f5b03f108917b71188/pkg/server/admin.go#L4104)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I don't think this is ok to do. `admin.go` can't pull in ccl code like this because it should be buildable as part of the OSS build.

See for instance how this code https://github.com/cockroachdb/cockroach/blob/a5ca1201e165d288ff188a6820843366028a920d/pkg/server/tenant.go#L828 works when CCL isn't present. Look at how the factory has a non-CCL implementation in a public `var` that gets overridden https://github.com/cockroachdb/cockroach/blob/50413ad4e7b87d4223f815858ae4655b179ad440/pkg/kv/kvclient/kvtenant/connector.go#L124

</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 4133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLajaUy4v8VT63WoL5Y:-NM-Nx1H-bD9bUPrPHK1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d35a2c3931172db4e60865f5b03f108917b71188/pkg/server/admin.go#L4133)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

instead of casting like this, consider implementing a method on the `onDemandServer` interface that gives you what you want. Then you can just call it in the loop without worrying about the tenant's special case.
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 4118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLeQwFn3J_jxqX_sbpy:-NM-NuTdE49yLalt1-z9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5e3eaf30740f5dd2662199d8f06ffb210f55e9/pkg/server/admin.go#L4118)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Please add `WHERE ... AND name IS NOT NULL` above and get rid of this.
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 4129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLeR6WQ0RIzEgNai9d1:-NM-Nv5qC8uUVxZPkkaI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f5e3eaf30740f5dd2662199d8f06ffb210f55e9/pkg/server/admin.go#L4129)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Consider logging the error + skipping this server (or add an error field for this tenant name in the response payload), instead of aborting the list entirely. Remember that the zip logic was meant to do best effort collection - it's ok if it skips over things that don't work.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-:-NM-v1laEe-BNoyiM0ZX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian)

___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NLaiWnL3-l4a742E6Nw):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Ack. Nothing actionable re: this PR, right?
</blockquote></details>

correct.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just need to move the test into `pkg/ccl` as I describe in comment.

> Reviewed 3 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-:-NM1AzOF8GjmG8d-4qha:b-1kkwan)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi and @knz)

___
*[`pkg/server/admin_test.go` line 75 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95024#-NM18ESu4R_nl33wwO6Z:-NM18ESu4R_nl33wwO6_:b-lo5sxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/93bf93b814fadcf5dd57b74e6b6a8e95a7694164/pkg/server/admin_test.go#L75)):*
> ```Go
> 
> // Dummy import to pull in kvtenantccl. This allows us to start tenants.
> var _ = kvtenantccl.Connector{}
> ```

I don't think you can pull this in within `pkg/server` either. I don't see existing usages like this. Move the `TestListTenants` test into `pkg/ccl/serverccl/adminccl/tenant_admin_test.go` instead.


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
TFTR!

bors r+
